### ISSUE TITLE
Rhino fails on jshint@2.7.0

### ISSUE THREAD

**Issue by segrey** [jshint/jshint#2308]
Rhino isn't working with jshint@2.7.0

```
$ java -jar ./js.jar -debug ./jshint-2.7.0.js
js: "jshint-2.7.0.js", line 1975: uncaught JavaScript runtime exception: TypeError: Cannot read property "prototype" from undefined
	at jshint-2.7.0.js:1975 (runInContext)
	at jshint-2.7.0.js:8270 (anonymous)
	at jshint-2.7.0.js:1525 (anonymous)
	at jshint-2.7.0.js:1515 (anonymous)
	at jshint-2.7.0.js:6 (s)
	at jshint-2.7.0.js:6 (anonymous)
	at jshint-2.7.0.js:12384 (anonymous)
	at jshint-2.7.0.js:6 (s)
	at jshint-2.7.0.js:17963 (anonymous)
	at jshint-2.7.0.js:4
```

```
$ java -version
java version "1.8.0_40"
Java(TM) SE Runtime Environment (build 1.8.0_40-b25)
Java HotSpot(TM) Server VM (build 25.40-b25, mixed mode)
```

Reproduced on [rhino-js-1_7R4](https://developer.mozilla.org/ru/docs/Rhino/Download_Rhino) and [rhino-js-1_7R5](https://github.com/mozilla/rhino/releases/tag/Rhino1_7R5_RELEASE): js.jar could be found inside rhino-js installation.

File jshint-2.7.0.js:  https://raw.githubusercontent.com/jshint/jshint/2.7.0/dist/jshint.js
File jshint-2.6.3.js:  https://raw.githubusercontent.com/jshint/jshint/2.6.3/dist/jshint.js

However rhino works fine for jshint@2.6.3 (empty output):
```
$java -jar rhino1_7R5/js.jar -debug ./jshint-2.6.3.js
$
```

**Comment by rwaldron** [comment]
> File jshint-2.7.0.js: https://raw.githubusercontent.com/jshint/jshint/2.7.0/dist/jshint.js
> File jshint-2.6.3.js: https://raw.githubusercontent.com/jshint/jshint/2.6.3/dist/jshint.js

That file isn't for Rhino, this one is: https://github.com/jshint/jshint/blob/master/dist/jshint-rhino.js

**Comment by segrey** [comment]
Yep, I know, but jshint.js can also be run with rhino interpreter, because jshint-rhino.js is just a [concatenation](https://github.com/jshint/jshint/blob/master/bin/build#L34) of jshint.js and [rhino cli runner](https://github.com/jshint/jshint/blob/master/src/platforms/rhino.js). My use-case is to run jshint inside java process communicating via rhino java API. The benefit is performance: initial jshint.js rhino-compilation is quite lengthy, but once compiled JSHINT function can be reused and it runs very fast. So, jshint-rhino.js can't be applied for my use-case, because it is solely intended for cli usage.

Anyway, the issue also affects [jshint-rhino-2.7.0.js](https://raw.githubusercontent.com/jshint/jshint/2.7.0/dist/jshint-rhino.js):
```
$ java -jar ./js.jar -debug ./jshint-rhino-2.7.0.js 
js: "./jshint-rhino-2.7.0.js", line 1977: uncaught JavaScript runtime exception: TypeError: Cannot read property "prototype" from undefined
	at ./jshint-rhino-2.7.0.js:1977 (runInContext)
	at ./jshint-rhino-2.7.0.js:8272 (anonymous)
	at ./jshint-rhino-2.7.0.js:1527 (anonymous)
	at ./jshint-rhino-2.7.0.js:1517 (anonymous)
	at ./jshint-rhino-2.7.0.js:8 (s)
	at ./jshint-rhino-2.7.0.js:8 (anonymous)
	at ./jshint-rhino-2.7.0.js:12386 (anonymous)
	at ./jshint-rhino-2.7.0.js:8 (s)
	at ./jshint-rhino-2.7.0.js:17965 (anonymous)
	at ./jshint-rhino-2.7.0.js:6
```

**Comment by rwaldron** [comment]
This isn't a bug in jshint. The code comes from the lodash source that's bundled in: https://github.com/jshint/jshint/blob/master/dist/jshint-rhino.js#L1977

@jdalton do you know anything about this?

**Comment by jdalton** [comment]
> do you know anything about this?

Try modifying the jshint package to reference `lodash-compat` instead of `lodash` to see if that helps with your Rhino issue. lodash is tested against rhino per commit. If the error persists I have a few other ideas.

**Comment by segrey** [comment]
It isn't a bug in lodash as it might seem (https://github.com/lodash/lodash/issues/365). I've tested: rhino works fine with lodash v2.4.1 and lodash v3.6.x.

The problem is in the browserified version of lodash.
[This lodash wrapper function](https://github.com/jshint/jshint/blob/master/dist/jshint-rhino.js#L1517) is invoked with `global` parameter set to `{}`.
Argument `global` is calculated [here](https://github.com/jshint/jshint/blob/master/dist/jshint-rhino.js#L8304):
```
typeof global !== "undefined" ? global : typeof self !== "undefined" ? self : typeof window !== "undefined" ? window : {}
```

In case of rhino `global`, `self` and `window` global variables are undefined.

**Comment by jdalton** [comment]
Browserify will usually create the `global` object if its referenced. Are you running browserify with options to ignore `global`?

**Comment by segrey** [comment]
@jdalton As I can see, no particular options used to ignore `global`. Here is a build script: https://github.com/jshint/jshint/blob/master/bin/build

**Comment by jdalton** [comment]
Seems like a Browserify bug. Try being explicit with it with the codified version of Browserify's `--insert-globals` flag or `--insert-global-vars` flag.

**Comment by rwaldron** [comment]
@segrey @jdalton thanks, I'll try these suggestions.

**Comment by segrey** [comment]
I can be wrong, but, AFAIK, rhino global variable is just `this`. That would probably complicate building correct string as `opts.insertGlobalVars` param. Well, I couldn't so far.
JFYI: I've come with another fix https://github.com/segrey/jshint/commit/1824d740c2d6656e663de43cf6d95fc8a642c796. Or it's better to fix this issue using browserify API?

**Comment by jdalton** [comment]
The empty object in [the related patch](https://github.com/segrey/jshint/commit/1824d740c2d6656e663de43cf6d95fc8a642c796#diff-81616a3b17429a2d7a410eca8a567de8L25) does look related to your [previous comment](https://github.com/jshint/jshint/issues/2308#issuecomment-92464768):
> This lodash wrapper function is invoked with `global` parameter set to `{}`.

lodash detection of variables is [trying to work well in multiple environments](https://github.com/lodash/lodash/blob/3.6.0/lodash.js#L253-L277) though browserify+rhino is something we haven't tried. It goes with a free var `global` first as browserify or r.js may create one, then `window`, then `self` for web workers, then `this`.

**Comment by segrey** [comment]
>The empty object in the related patch does look related to your previous comment:

>>This lodash wrapper function is invoked with global parameter set to {}.

Indeed, if undefined, `window` shouldn't be set to `{}`. Otherwise, it might influence the value of passed `global` parameter. Moreover, setting a free var `window` just doesn't make sense for rhino environment. Actually, setting a free var `global` doesn't make sense for rhino environment either, but it allows to fix the issue. Yep, hacky way ;)

**Comment by DanielRuf** [comment]
I have the same problem. jshint 2.7.0 does not work anymore.

**Comment by segrey** [comment]
@DanielRuf As a workaround, you can use this patched build used by JetBrains IntelliJ IDEs: http://download.jetbrains.com/idea/jshint/jshint-2.7.0.js

**Comment by DanielRuf** [comment]
I am using JSHint 2.6.3 now which works without any problems. Just wanted to let you know that this is a general problem which possibly affects everyone who uses JSHint 2.7.0.

**Comment by segrey** [comment]
Rhino still fails on jshint@2.8.0.

**Comment by lukeapage** [comment]
@segrey if you make a PR, it can be considered - I'm not in a position to review rhino stuff though, I don't use it.
It seems a bit nasty to me, to have rhino code inside the js bootstrapper, but then we are also doing `if (typeof window === 'undefined') window = {};` which is possibly just as bad :S
Why are we setting window to {}?

**Comment by DanielRuf** [comment]
@lukeapage to define the window object, see here for more details: https://github.com/jshint/jshint/issues/1100

**Comment by busches** [comment]
@segrey are you still planning on opening a PR for this? Looks like someone else has also submitted a fix, https://github.com/jshint/jshint/pull/2774, but needs help with tests backing it up.

**Comment by cscott** [comment]
Confirming that the rhino build broke between jshint 2.6.3 and 2.7.0 and remains broken in the latest version of jshint (2.9.1).  I suggest at least updating the docs to indicate that rhino support is known to be broken.

#2774 doesn't look like a real fix to me, just a hack.

https://github.com/segrey/jshint/commit/1824d740c2d6656e663de43cf6d95fc8a642c796 looks pretty good to me, though.  That should become a PR.

**Comment by jdalton** [comment]
Lodash has bumped to v4. Might be worth seeing if an upgrade makes this issue go away.

**Comment by g-gabber** [comment]
Rhino is still broken @ 2.9.2

**Comment by kived** [comment]
I'm using V8, not Rhino, but I'm getting the same error with 2.9.3 (tried both jshint.js and jshint-rhino.js after removing the `#!`). The "patched" 2.7.0 build linked above didn't work for me either.

2.6.3 works great though.

**Comment by alexgman** [comment]
Is there a linter available for use with Rhino 1.7.7c ?
